client_plugin/drivers/shared/etcdops.go

```diff
@@ -0,0 +1,271 @@
+// Copyright 2017 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package shared
+
+import (
+	"context"
+	"fmt"
+	"os/exec"
+	"strings"
+
+	log "github.com/Sirupsen/logrus"
+	etcdClient "github.com/coreos/etcd/clientv3"
+	"github.com/docker/engine-api/client"
+	"github.com/docker/engine-api/types"
+	"github.com/docker/engine-api/types/swarm"
+)
+
+const (
+	apiVersion               = "v1.24" // docker engine 1.24 and above support this api version
+	dockerUSocket            = "unix:///var/run/docker.sock"
+	etcdClientPort           = ":2379"
+	etcdPeerPort             = ":2380"
+	etcdClusterToken         = "vsphere-shared-etcd-cluster"
+	etcdListenURL            = "0.0.0.0"
+	etcdScheme               = "http://"
+	etcdClusterStateNew      = "new"
+	etcdClusterStateExisting = "existing"
+)
+
+// InitEtcd start or join ETCD cluster depending on the role of the node
+func InitEtcd() error {
+	// create new docker client
+	cli, err := client.NewClient(dockerUSocket, apiVersion, nil, nil)
+	if err != nil {
+		log.WithFields(
+			log.Fields{"error": err},
+		).Error("Failed to create client for Docker ")
+		return err
+	}
+
+	ctx := context.Background()
+
+	// get NodeID from docker client
+	info, err := cli.Info(ctx)
+	if err != nil {
+		log.WithFields(
+			log.Fields{"error": err},
+		).Error("Failed to get Info from docker client ")
+		return err
+	}
+
+	// worker just returns
+	nodeID := info.Swarm.NodeID
+	if info.Swarm.ControlAvailable == false {
+		log.WithFields(
+			log.Fields{"nodeID": nodeID},
+		).Info("Swarm node role: worker, return from InitEtcd ")
+		return nil
+	}
+
+	// check my local role
+	node, _, err := cli.NodeInspectWithRaw(ctx, nodeID)
+	if err != nil {
+		log.WithFields(log.Fields{"nodeID": nodeID,
+			"error": err}).Error("Failted to inspect node ")
+		return err
+	}
+
+	// get the IP address of current node
```

Comment:
Yes for here we can get NodeAddr from Info.Swarm.NodeAddr.
However later when we need to get the address of leader from a non-leader manager,
we cannot do "docker info" for a remote node, but can only do "docker node inspect".
The returned struct of NodeInspectWithRaw is swarm.Node structure, instead of types.Info.
So we have to go through swarm.Node.ManagerStatus.Addr to get the addresses...
And then I decide to reuse that code :)